cadcdata/cadcdata/core.py

```diff
@@ -0,0 +1,418 @@
+        resource_url = urlparse(resource_id)
+        self.host = host
+
+        agent = "cadc-data-client/{}".format(version.version)
+
+        self._repo_client = net.BaseWsClient(resource_id, subject,
+                                             agent, retry=True, host=self.host)
+        self.logger.info('Service URL: {}'.format(self._repo_client.base_url))
+
+
+    def get_file(self, archive, file_id, file=None, decompress=False,
+                 cutout=None, fhead=False, wcs=False, process_bytes=None):
+        """
+        Get a file from an archive. The entire file is delivered unless the cutout argument
+         is present specifying a cutout to extract from the file.
+        :param archive: name of the archive containing the file
+        :param file_id: the ID of the file to retrieve
+        :param file: file to save data to (file or file_name). If None, it uses the response
+        content disposition to determine the name of the destination file
+        :param cutout: perform cutout operation on the file before the result is delivered
+        :param wcs: True if the wcs is to be included with the file
+        :param process_bytes: function to be applied to the received bytes
+        :return: the data stream object
+        """
+        assert archive is not None
+        assert file_id is not None
+        resource = '/{}/{}'.format(archive, file_id)
+        params = {}
+        if fhead:
+            params['fhead'] = fhead
+        if wcs:
+            params['wcs'] = wcs
+        if cutout:
+            params['cutout'] = cutout
+        self.logger.debug('GET '.format(resource))
+        response = self._repo_client.get(resource, stream=True, params=params)
+        if file is not None:
+            if not hasattr(file, 'read'):
+                # got a file name?
+                with open(file, 'w') as f:
+                    self._save_bytes(response, f, decompress, process_bytes)
+            else:
+                self._save_bytes(response, file, decompress, process_bytes)
+        else:
+            # get the file name from the content disposition
+            content_disp = response.headers.get('content-disposition', '')
+            file = file_id
+            for content in content_disp.split():
+                if 'filename=' in content:
+                    file = content[9:]
+            if file.endswith('gz') and decompress:
+                file = os.path.splitext(file)[0]
+            self.logger.debug('Using content disposition file name: {}'.format(file))
+            with open(file, 'w') as f:
+                self._save_bytes(response, f, decompress, process_bytes)
+        self.logger.debug('Successfully saved file\n')
+
+
+    def _save_bytes(self, response, file, decompress=False, process_bytes=None):
+        # requests automatically decompresses the data. Tell it to do it only if it had to
+        if not decompress:
```

Comment:
if not decompress:
   reader = response.raw.read
else:
   reader = response.iter_content

then call 'reader(READ_BLOCK_SIZE)'  an just have a single loop. instead of re-producing the code.